### PR TITLE
Cyclonedx support

### PR DESCRIPTION
**Pull request by nadgowdas**
Added Support for Processor/Parser/Ingestor for cycloneDX SBOM

### CODE REVIEW

File: docs/alpine-cyclonedx.json [hunk 1]
```diff
@@ -0,0 +1,1211 @@
+{
```
Comment (lumjjb) [inline docs/alpine-cyclonedx.json:1]: did you mean to have this folder in the PR as well? Or an accidental local upload?

File: pkg/handler/processor/cyclonedx/cyclonedx.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+// CycloneDXProcessor processes CycloneDXProcessor documents.
+// Currently only supports CycloneDX-JSON documents
+type CycloneDXProcessor struct {
+}
+
+func (p *CycloneDXProcessor) ValidateSchema(d *processor.Document) error {
+	if d.Type != processor.DocumentCycloneDX {
+		return fmt.Errorf("expected document type: %v, actual document type: %v", processor.DocumentSPDX, d.Type)
```
Comment (lumjjb) [inline pkg/handler/processor/cyclonedx/cyclonedx.go:18]: 's/SPDX/CycloneDX/'

File: pkg/handler/processor/guesser/type_guesser.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ func init() {
 	_ = RegisterDocumentTypeGuesser(&ite6TypeGuesser{}, "ite6")
 	_ = RegisterDocumentTypeGuesser(&dsseTypeGuesser{}, "dsse")
 	_ = RegisterDocumentTypeGuesser(&spdxTypeGuesser{}, "spdx")
+	_ = RegisterDocumentTypeGuesser(&cycloneDXTypeGuesser{}, "cyclonedx")
```
Comment (lumjjb) [inline pkg/handler/processor/guesser/type_guesser.go:28]: can we add a test case to `type_guesser_test.go` as well to include a CDX document?

File: pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	}
+}
+
+func (c *cyclonedxParser) CreateNodes(ctx context.Context) []assembler.GuacNode {
+	nodes := []assembler.GuacNode{}
+	addNodes(c.rootPackage, &nodes)
+	return nodes
+}
+
+func addNodes(curPkg parentPackages, nodes *[]assembler.GuacNode) {
```
Comment (lumjjb) [inline pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:35]: just a slice instead? the slice is already a pointer type

File: pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func (c *cyclonedxParser) addRootPackage(cdxBom *cdx.BOM) {
+	// oci purl: pkg:oci/debian@sha256%3A244fd47e07d10?repository_url=ghcr.io/debian&tag=bullseye
+	splitImage := strings.Split(cdxBom.Metadata.Component.Name, "/")
+	if len(splitImage) == 3 {
+		rootPackage := assembler.PackageNode{}
+		rootPackage.Purl = "pkg:oci/" + splitImage[2] + "?repository_url=" + splitImage[0] + "/" + splitImage[1]
+		rootPackage.Name = cdxBom.Metadata.Component.Name
+		rootPackage.Digest = append(rootPackage.Digest, cdxBom.Metadata.Component.Version)
```
Comment (lumjjb) [inline pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:85]: quick question: is digest in the format like "sha256:abc..."?

File: pkg/ingestor/parser/parser.go [hunk 6]
```diff
@@ -95,6 +96,14 @@ func parseHelper(ctx context.Context, doc *processor.Document) (*common.GraphBui
 		}
 		slsaGraphBuilder := common.NewGenericGraphBuilder(slsaParser, nil)
 		return slsaGraphBuilder, nil
+	case processor.DocumentCycloneDX:
+		cyclonedxParser := cyclonedx.NewCycloneDXParser()
+		err := cyclonedxParser.Parse(ctx, doc)
+		if err != nil {
+			return nil, err
+		}
+		slsaGraphBuilder := common.NewGenericGraphBuilder(cyclonedxParser, nil)
```
Comment (lumjjb) [inline pkg/ingestor/parser/parser.go:105]: nit: rename from slsa

**Review by lumjjb — COMMENTED** [review]
minor nits - else lgtm

File: pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go (hunk 5 above)
Comment (nadgowdas) [reply pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:85]: Yes, snippet from bom:
```
"component": {
      "bom-ref": "685d54d5e8f71d2f",
      "type": "container",
      "name": "docker.io/library/alpine:latest",
      "version": "sha256:cbaa390ec61abd889dfa5146a411c11717c19ef9e5e18458dbdc06439076ffe7"
    }
```

File: pkg/ingestor/parser/parser.go (hunk 6 above)
Comment (nadgowdas) [reply pkg/ingestor/parser/parser.go:105]: done!

File: pkg/handler/processor/cyclonedx/cyclonedx.go (hunk 2 above)
Comment (nadgowdas) [reply pkg/handler/processor/cyclonedx/cyclonedx.go:18]: done!

File: docs/alpine-cyclonedx.json (hunk 1 above)
Comment (nadgowdas) [reply docs/alpine-cyclonedx.json:1]: ahh my bad.. will remove this

File: pkg/handler/processor/guesser/type_guesser.go (hunk 3 above)
Comment (nadgowdas) [reply pkg/handler/processor/guesser/type_guesser.go:28]: done!

File: pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go (hunk 4 above)
Comment (nadgowdas) [reply pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:35]: hmm..but we can't append with slice ?

Comment (lumjjb) [reply pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go:50]: yea another pattern is returning the slice, append does add it to the current list most of the time but the language doesn't guarantee it. This is fine.

**Review by lumjjb — APPROVED** [review]
LGTM

File: pkg/ingestor/parser/parser.go [hunk 7]
```diff
@@ -95,6 +96,14 @@ func parseHelper(ctx context.Context, doc *processor.Document) (*common.GraphBui
 		}
 		slsaGraphBuilder := common.NewGenericGraphBuilder(slsaParser, nil)
 		return slsaGraphBuilder, nil
+	case processor.DocumentCycloneDX:
```
Comment (pxp928) [inline pkg/ingestor/parser/parser.go:99]: add a test case for cycloneDX in `parser_test.go`

Comment (pxp928) [reply pkg/ingestor/parser/parser.go:99]: no longer needed.

**Review by pxp928 — APPROVED** [review]
LGTM

**Pull request merged**